### PR TITLE
Prevent starting a VM in destroyed state (or any state but Stopped)

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

A destroyed VM when attempted to be started (using the bulk action support in the UI), results in a successful operation, however, the VM doesn't actually start. This PR fixes the false positive result shown in the above mentioned condition

image

VM continues to remain in destroyed state - but a success response is returned

image

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [X] Trivial

### How Has This Been Tested?
Tried starting an instance in destroyed state:
```
2021-06-29 12:57:51,609 WARN  [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-5:ctx-da4601fa job-52/job-54 ctx-cc08e96e) (logid:08e57fcb) VM VM instance {"id": "3", "name": "i-2-3-VM", "uuid": "4213c3d9-72a8-4201-96c4-003e2c2f845b", "type"="User"} is not in a state to be started: Destroyed
2021-06-29 12:57:51,610 INFO  [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-5:ctx-da4601fa job-52/job-54 ctx-cc08e96e) (logid:08e57fcb) Caught CloudRuntimeException, returning job failed com.cloud.utils.exception.CloudRuntimeException: Cannot start VM: VM instance {"id": "3", "name": "i-2-3-VM", "uuid": "4213c3d9-72a8-4201-96c4-003e2c2f845b", "type"="User"} in Destroyed state
2021-06-29 12:57:51,613 DEBUG [c.c.v.VmWorkJobHandlerProxy] (Work-Job-Executor-5:ctx-da4601fa job-52/job-54 ctx-cc08e96e) (logid:08e57fcb) Done executing VM work job: com.cloud.vm.VmWorkStart{"dcId":0,"rawParams":{"VmPassword":"rO0ABXQADnNhdmVkX3Bhc3N3b3Jk"},"userId":2,"accountId":2,"vmId":3,"handlerName":"VirtualMachineManagerImpl"}

```
image

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -881,8 +881,8 @@ protected boolean checkWorkItems(final VMInstanceVO vm, final State state) throw
             }
 
             if (state != State.Stopped) {
-                s_logger.debug("VM " + vm + " is not in a state to be started: " + state);
-                return null;
+                s_logger.warn("VM " + vm + " is not in a state to be started: " + state);
+                throw new CloudRuntimeException(String.format("Cannot start VM: %s in %s state", vm, state));
```
Comment (GutoVeronezi) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:885]: We could unify these messages in a String var (with String.format).

Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:885]:
```suggestion
                String msg = String.format("Cannot start VM: %s in %s state", vm, state)
                s_logger.warn(msg);
                throw new CloudRuntimeException(msg));
```

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
cltgm

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 2]
```diff
@@ -881,8 +881,9 @@ protected boolean checkWorkItems(final VMInstanceVO vm, final State state) throw
             }
 
             if (state != State.Stopped) {
-                s_logger.debug("VM " + vm + " is not in a state to be started: " + state);
-                return null;
+                String msg = String.format("Cannot start VM: %s in %s state", vm, state);
```
Comment (GutoVeronezi) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:884]: `vm` is of type `VMInstanceVO`; This class implements a `toString()` like `VM instance...`, so the final message will be:

`Cannot start VM: VM instance...`.

We could remove the `VM:` here.

**Review by GutoVeronezi — APPROVED** [review]
CLGTM; I raised a last point about `toString`.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 2 above)
Comment (Pearl1594) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: keen observation

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 3]
```diff
@@ -881,8 +881,9 @@ protected boolean checkWorkItems(final VMInstanceVO vm, final State state) throw
             }
 
             if (state != State.Stopped) {
-                s_logger.debug("VM " + vm + " is not in a state to be started: " + state);
-                return null;
+                String msg = String.format("Cannot start %s in %s state", vm, state);
+                s_logger.warn(msg);
+                throw new CloudRuntimeException(msg);
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:886]: code LGTM

**Review by sureshanaparti — APPROVED** [review]

**Review by vladimirpetrov — APPROVED** [review]
Functionality tested - LGTM.

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

**Comment by sureshanaparti** [comment]
Merging this based on LGTMs, manual and smoke tests.

**Pull request merged**